tests/smtlib/testfile-steps-bound.dolmen.smt2

```diff
@@ -0,0 +1,22 @@
+(set-logic ALL)
+
+(declare-const x Int)
+(declare-const y Int)
+(assert (= (* x x) 3))
+
+; Should answer 'unknown'
+(push 1)
+(set-option :steps-bound 2)
```

Comment:
I am not sure it is worth being able to define the `set-option` after stuff has been defined/pushed/etc. This is an option that should be supported at the top-level (i.e. before any constants/asserts) but we really don't need to support it past that — is there an use case here?

Reply:
Not a use case, but I did find a bug while implementing this when setting an option after a push, so I think we should keep the test as it is.

Reply:
I don't think we should have tests for features that we don't want to support, unless if the tests says that it should be an error. Otherwise, that is implicitly saying that we support the feature.

Reply:
Let's say that I want to support this feature then, the whole test is a use case (let's try something with a given limit, then with a greater limit).

Reply:
Besides, we don't want alt-ergo to throw random `assert false` even if the user writes non-sense in its smt file

Reply:
You are making a false dichotomy in your second message. The options are not "support this use case" or "raise an assertion failure"; there is a mechanism for reporting proper errors to the user that can be used. If we currently raise an assertion failure for this test case, that is a bug, what we should do is display an appropriate error instead.

I don't really understand the point of setting a limit then a greater limit in the current infrastructure, since the solver will have to re-do everything it did the second time, so it seems to me you'd just be wasting time. If the internal solver infrastructure was able to support this and we could say "ignore the previously set limit and keep solving", why not, I'd be happy to throw that in. But currently, we can't; if the limit is reached, we have to start from scratch, which induces additional complexity (need to reset the solver, replay the assertion queue, make sure everything is properly synced, etc.) of the kind you just advocated for removing yesterday.

In any case: this is currently not supported (as evidenced by the fact it is neither documented nor working), and it would require committing to additional complexity to support properly given the current infrastructure (again, because we can't just re-use the solver). This doesn't sound like a decision that should be taken by introducing a test case in a PR; it sounds like a change that an issue should be created for and evaluated separately. 

Given that it is also a thing I have never needed when using solvers, nor heard anyone requesting — to my knowledge, resource limits are typically set once at the beginning and that's it — I would be interested if you have practical use cases for changing limits during execution (besides "just because"), but that discussion should be happening in a separate issue if it has to happen, not holding up the merge of this PR.

Reply:
> You are making a false dichotomy in your second message.
 
Sorry if I was not clear. I meant to say two separate things: that *I* did want this feature (because I'm tired of arguing for that, it is a *test*) ; and separately, that I didn't want `assert false` when setting an option between a push and a pop (which was a real issue when implementing, that this PR fixed). 

It is not even like push and pop had any effect on the matter, push and pop only affect the assertions. IMO, the simple fact that it triggered an assertion at some point may imply that we lack tests that push alt-ergo to its edge.

So, if you *really* want this line gone, can we at least agree on creating another test file that presses alt-ergo on strange cases? I would be fine keeping the line for this exact same reason.

Reply:
> It is not even like push and pop had any effect on the matter, push and pop only affect the assertions. IMO, the simple fact that it triggered an assertion at some point may imply that we lack tests that push alt-ergo to its edge.

This is definitely 100% true, and somewhere where we have to improve! Not arguing with that at all.

> So, if you really want this line gone, can we at least agree on creating another test file that presses alt-ergo on strange cases? I would be fine keeping the line for this exact same reason.

(Note, I don't want that line gone — I want that line moved before the `push`)

I would be happy with having tests that explore corner cases and other oddities, that would be a good thing. What I am *not* happy with, however, is tests that test an unsupported feature (except to test that it fails with a proper error message, of course). Doing so enshrines the behavior, so that it effectively becomes supported: next time someone makes a change, the test will fail, and they will dutifully make sure that it still passes in the same way. That's the point of having tests, after all. I felt that I made clear that this was my stance, but I think this was partly in the meeting last yesterday and not in this discussion, so I apologize if there was confusion.

Now, here, given the current infrastructure and limitation regarding re-starting the solver after the step limit is reached, the feature set that we can safely commit to support in a coherent and satisfactory way is the use of `(set-option :step-bounds)` in `init` mode (i.e. before any `push`, `pop`, `assert`, etc.). Well, I guess that is not entirely true: we can support the use of `(set-option :step-bounds)` *as long as the step limit is not reached*, and then we can no longer allow increasing the step limit without additional infrastructure -- but that would make the validity of the command depend on internal Alt-Ergo details which is not good user experience.

Given that fact, and given the stance above, the tests should either be in that fragment, or if they are outside of that fragment they should either be errors or contain comments that say "hey we don't do the right thing here currently, it's OK if the test outcome change". But ideally such tests should be kept to a minimum, because it encourages a lot of churn in expected files which leads to paying less attention to changes in test outputs that can be wrong.

I am sorry to be arguing so much on this test, but hopefully my position makes sense here.

I think the options I would be happy with are:
 - Change the test so that the `(set-option)` are in init mode (i.e. before any `push`, `pop`, `declare`, `assert`, etc)
 - Keep the test as is, but change the code so that it *actually fails* instead of silently doing something (we already use an "are we in init mode" test for the "sat-solver" option, it can be reused from there)
 - Keep the test as is, but explicitly add a comment in the test file saying "this is unsupported but we want to make sure we don't crash".

Reply:
> Keep the test as is, but change the code so that it actually fails instead of silently doing something (we already use an "are we in init mode" test for the "sat-solver" option, it can be reused from there)

I am OK with this solution, but I think this belongs to another PR (that would slightly change the d_solver_options module).

Reply:
By the way, this test is expected to fail in a near future; the first assertion requires mores than 2 steps to be completed, but as it is the option *is* set at initialization because we do not treat SMT instructions imperatively. When we do, this test should fail.
I will add this fact to the test in case I am not assigned to the specific PR that will trigger the error.

Reply:
Thank you, I am OK with this solution. I will approve the changes as soon as the info is added to the test case.

Can you point to the corresponding PR/issue related to making sure the test fails for tracking purposes, I am not sure which one it is.

Reply:
Maybe something linked to #899 ? The issue is definitely linked to what we discussed on tuesday about the imperative execution of smtlib instructions.